Fixes #6835 - support filtered API for oVirt

ares opened the pull request.

ares committed:
Fixes #6835 - support filtered API for oVirt

ohadlevy commented on db/migrate/20160308100307_add_filtered_api_to_ovirt.rb:
```diff
@@ -0,0 +1,5 @@
+class AddFilteredApiToOvirt < ActiveRecord::Migration
+  def change
+    add_column :compute_resources, :filtered_api, :boolean, :default => false, :null => false
```
you can use the serialized attribute instead of adding db columns that do not fit other compute resources?

ares replied:
I can, I don't want to :-) especially in this table I don't expect thousands of records and I think the number of downsides it brings is too big (migration touching serialized data, possible inconsistency between two objects of same class, faulty implementation that uses method_missing without defining respond_to?)

ohadlevy commented on app/models/compute_resources/foreman/model/ovirt.rb:
```diff
@@ -250,7 +250,8 @@ def client
           :ovirt_datacenter => uuid,
-          :ovirt_ca_cert_store => ca_cert_store(public_key)
+          :ovirt_ca_cert_store => ca_cert_store(public_key),
+          :ovirt_filtered_api => filtered_api
```
does that require a minimum version of rbovirt / ovirt it self?

ares replied:
this was introduced in fog 1.28, we are already on [1.37](49569b06bd38e508078f9e1257467da13596ef00), rbovirt seems to support filtered API since [v0.0.24](https://github.com/abenari/rbovirt/commit/494028e14948d89ab331e79d6e64fdfeec5bae4c) and were using [0.0.32+](https://github.com/abenari/rbovirt/commit/494028e14948d89ab331e79d6e64fdfeec5bae4c) not sure about when they added it in oVirt but I guess if users oVirt does not support it, they can simply keep it disabled (I think it would be very old version)

ares commented:
@ohadlevy anything else I could do to get this ready for merge?

iNecas commented:
Tested and works well. :+1:  code-wise from me

ares commented:
Thank you, setting ready to merge so someone with permissions could merge.

domcleal commented:
Sorry, I'm going to unset it unless a committer's reviewed it.  (Looks like that's @ohadlevy.)

ohadlevy commented:
/me delegate to @tbrisker :)

ares commented:
Great, it got things moving :-) Sorry if it was too bold.

tbrisker commented on app/views/compute_resources/form/_ovirt.html.erb:
```diff
@@ -1,6 +1,7 @@
 <%= text_f f, :url, :size => "col-md-8", :help_block => _("e.g. https://ovirt.example.com/api"), :help_inline => documentation_button('5.2.7oVirt/RHEVNotes') %>
 <%= text_f f, :user, :help_block => _("e.g. admin@internal") %>
 <%= password_f f, :password, :unset => unset_password? %>
+<%= checkbox_f f, :filtered_api, :help_block => _("If you don't have admin permissions you can use slower filtered API, that works with user account."), :label => _('Filtered API') %>
```
How about making the `help_block` an inline popover?
Also, nitpick, you are repeating yourself in the text. I think `"If you don't have admin permissions you can use slower filtered API."` should be enough.

ares replied:
that was my first attempt but it looked weird
![popover](https://cloud.githubusercontent.com/assets/109773/13751014/5e0978a0-ea08-11e5-8faa-aa831684ef93.png)

but I can add it if you think it's better

tbrisker replied:
I think if you use `:help_inline` instead of `:help_block` it should look better.

ares replied:
I'm not sure if it looks any better
![popover1](https://cloud.githubusercontent.com/assets/109773/13773604/6ab79766-ea9a-11e5-8e34-5ad55cf1dfee.png)

This is what we have for interfaces flag, also does not look great to me.
![popover2](https://cloud.githubusercontent.com/assets/109773/13773701/f86f5ab2-ea9a-11e5-9a93-8ecd0bc085f4.png)

I think I find this the best.
![popover3](https://cloud.githubusercontent.com/assets/109773/13773730/252b2964-ea9b-11e5-887f-b259330132f9.png)

Which one do you prefer?

tbrisker commented:
@ares tested, this works just fine. Once the comment about the help text is fixed I'll merge.

ares commented:
I changed the help text, please let me know about the popover.

ares committed:
Fixes #6835 - improve error detection and help text

dLobatog commented on app/views/compute_resources/form/_ovirt.html.erb:
```diff
@@ -1,6 +1,7 @@
 <%= text_f f, :url, :size => "col-md-8", :help_block => _("e.g. https://ovirt.example.com/api"), :help_inline => documentation_button('5.2.7oVirt/RHEVNotes') %>
 <%= text_f f, :user, :help_block => _("e.g. admin@internal") %>
 <%= password_f f, :password, :unset => unset_password? %>
+<%= checkbox_f f, :filtered_api, :help_inline => _("If you don't have admin permissions you can use slower filtered API."), :label => _('User level API') %>
```
https://github.com/theforeman/foreman/pull/3308/files#r56128924

ares replied:
I believe that's already incorporated. I changed the label to "User level API" and it mentions that it can be used you user does not have administrator permissions.

lzap replied:
I think we are exposing too low-level stuff to the end-user. The checkbox should perhaps read `[x] Admin account` or something. The current name is too fuzzy and the inline help does not describe it enough. Why I should check it? What happens if I don't? I vote for _Administrator account_ with help text of something like "Faster filtering for priviledged accounts".

ares replied:
I don't see a big difference between "User level API" vs "Administrator account". Before I change it once again, could we have some poll to get the final text please? Just note that administrator account switches the meaning of flag so I should probably rename the field in DB as well, I guess for developer filtered API is more meaningful though.

lzap replied:
Don't you see the difference? Any user can perhaps tell if the account he/she just entered in the form was admin account. But User level API? What on Earch does it mean? I didn't know and had to google it today.

My helper text proposal tries to explain that pagination and searching is actually not possible.

lzap replied:
And for the record - I don't insist. This is just a comment, I don't test or review this one. I ended up on this page and had this on my mind. Up to you.

ares replied:
"Faster filtering for priviledged accounts" does not sound clear to me, not sure how it suggests anything about pagination and searching not being possible. I'm happy to change the inline help text but (anyone) please suggest the full sentence. Or more sentences that we'd hide in popover. I think the existing text is best so far. Maybe "User level API" -> "User account" label change would be enough?

tbrisker replied:
I would go with whatever terminology is used in the oVirt documentation. 
Also, its not only slower but also limited. I guess something like:
"Use the <oVirt term> API which is limited and slower but is available to non-admin users."

lzap replied:
> I would go with whatever terminology is used in the oVirt documentation. 

Out of curiosity, can you send a link to the doc that describes this
term? Because to me, it looks like there is no oVirt documentation at
all. ;-)

## 

Later,
 Lukas #lzap Zapletal

oourfali replied:
I can express again my thoughts on not having this support... I think it
shouldn't be added.
It will not be slower. not in every use-case.
I think it is weird to write that down at all.
It is not filtered API, it is our user-level API.
I think you're getting yourself into mess by adding this.

On Mon, Mar 21, 2016 at 10:45 AM, Lukáš Zapletal notifications@github.com
wrote:

> In app/views/compute_resources/form/_ovirt.html.erb
> https://github.com/theforeman/foreman/pull/3288#discussion_r56792403:
> 
> > @@ -1,6 +1,7 @@
> >  <%= text_f f, :url, :size => "col-md-8", :help_block => _("e.g. https://ovirt.example.com/api"), :help_inline => documentation_button('5.2.7oVirt/RHEVNotes') %>
> >  <%= text_f f, :user, :help_block => _("e.g. admin@internal") %>
> >  <%= password_f f, :password, :unset => unset_password? %>
> > +<%= checkbox_f f, :filtered_api, :help_inline => _("If you don't have admin permissions you can use slower filtered API."), :label => _('User level API') %>
> 
> I would go with whatever terminology is used in the oVirt documentation.
> Out of curiosity, can you send a link to the doc that describes this term?
> Because to me, it looks like there is no oVirt documentation at all. ;-)
> … <#-1306898184106590848_>
> -- Later, Lukas #lzap Zapletal
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/3288/files/a8b18e0690cea4dce35e07dc7a427bff0e072036#r56792403

ares commented:
@tbrisker I improved error handling that was asked in https://github.com/theforeman/foreman/pull/3308#discussion-diff-56128924. I keep it in separate commit for easier reviewing, I'll squash after review.

ares commented:
btw tests failures are unrelated

tbrisker commented on app/models/compute_resources/foreman/model/ovirt.rb:
```diff
@@ -86,17 +86,24 @@ def test_connection(options = {})
+          errors[:url] << _('HTTPS URL is required for API access')
+        when /insufficient permissions/
+          # ovirt does not add code 400 to error message
+          errors[:user] << e.message + ' ' + _('If your username and password is correct, try using user level API.')
```
Also make sure terminology here matches that on the checkbox.

ares commented:
@tbrisker I've modified texts. Sadly, it's not well documented in oVirt docs and it's always the same API just with different header. Anyway I think this should be clear enough now. I also added the value to v2 API. Please re-review.

tbrisker commented on app/views/compute_resources/form/_ovirt.html.erb:
```diff
@@ -1,6 +1,7 @@
 <%= text_f f, :url, :size => "col-md-8", :help_block => _("e.g. https://ovirt.example.com/api"), :help_inline => documentation_button('5.2.7oVirt/RHEVNotes') %>
 <%= text_f f, :user, :help_block => _("e.g. admin@internal") %>
 <%= password_f f, :password, :unset => unset_password? %>
+<%= checkbox_f f, :filtered_api, :help_inline => _("Use Filter header which makes API limited and slower but is available to non-admin users."), :label => _('User account') %>
```
Sorry to be nitpicking again on this, I think `"Use the 'Filter' header for API calls, which makes them limited and slower, but available to non-admin users."` would be clearer, with the label being `Filter API calls`. 
Also the help text now breaks a line and looks bad, please reconsider using a popover.

ohadlevy replied:
I  wonder if we should consider this a plugin vs something we want to
maintain in foreman itself, after talking to a view ovirt guys, they
recommended we do not use the filtered api at all.

iNecas replied:
No plugin please. This pretty much limits the usage of ovirt just to ones with admin rights. If something changes on ovirt side, we can reconsider, but right now, it's really solving real issue for us (and I expect others are in a situation they can't have admin account)

ares replied:
@tbrisker no need to applologize, I'm glad we're moving somewhere. Fixed.

@ohadlevy I'm not sure I properly understand your suggestion, create ovirt plugin that adds this checkbox? On our side it does not add much to maintain. If it does not work on ovirt side, I hope they are willing to fix it.

ohadlevy replied:
the problem is that it present a follow up problems, for example:
- if you are a large env, and has a lot of hosts, the vm list might not work
- associate vms will not work as well either (both listing the vms).

this also limit us to support better leverage ovirt, for example, we want
for associate vms to use search to find relevant mac addresses instead of
in memory searching for all macs, since search is not exposed, we would not
be able to use that.
I wonder if the better way will be to ask for an admin account - note that
in oVirt, admin does not have a full permissions (e.g. superuser) just
labeled as admin

I also don't like this, but I'm worried that a limitation on oVirt side
will create confusion for our users.

ohadlevy replied:
@ares I'm not aware they prioritize fixing this issue, maybe @oourfali can expand on it?

ares replied:
@ohadlevy adding such features is great but I'm not sure whether big setup (probably lower % of users) should cut support for feature that would work only with smaller setups (higher number of users I guess)

I can imagine introducing some capabilities model that would tell us what a given CR is capable of, it could be handy also to manage versions of CR (with different feature sets)

tbrisker commented:
This works and code & text are good to merge as far as I'm concerned once tests finish. 
However, we need to make a decision whether we want to make this change at all. 
I'm afraid I do not have enough experience with oVirt and I don't know enough about users' use cases to make that call, so I will leave that up to those of you who know more to make the decision.

oourfali commented:
There are many gaps in the filtered api. It wasn't designed for the use foreman does with the API. 
IMHO it shouldn't be pushed, as it might cause other issues. Even if things will work, they will be less performant.

ares commented:
@oourfali is there a different way to use ovirt API without admin account that foreman (or anything else) could use for automating the provisioning? There was some interest https://bugzilla.redhat.com/show_bug.cgi?id=1123676

iNecas commented:
@oourfali out next step is to add jsessionid support (we already have support on rbovirt side), from the performance reasons.

oourfali commented:
The performance aspects are different. Not related to the session id. 
The user api works differently, running queries rather than using the search mechanism. So, for different things might be less performant, especially when combined with search.

iNecas commented:
Regardless the performance limitations, I still think not providing the user-level accounts an ability to use Foreman with this credentials is a serious limitation. We have even customer cases already demanding this. If the ovirt has performance limitations on this use-case, I would rather see improvements on the ovirt side to improve this, rather then avoiding using it at all. If the user-level API was officially deprecated and communicated, that this will not be supported anymore (and the question would be when, because if it's too far away, it still might be worth having it here until the day comes).

If there are use-cases that the filtered API is not good for, in ideal world the API would be explicit enough about that, when this use comes. On the other hand, I can't think of more basic use-case for a virtualization platform than provisioning a host. I can imagine limiting the other cases (such as listing VMs) not being available in case the calls are deprecated/show serious issues. But the basic end-to-end provisioning via API seems to be valid use-case that I would expect from ovirt to handle without need for admin-level account.

From my perspective, this is pretty small change, that unblocks non-admin users using it and can be reverted/disable anytime an Ovirt version is released that deprecates it.

What about a note about about the lower performance in the description?

oourfali commented:
Foreman is in your hands... We support the user-level API completely, but some things work differently, and some functionality doesn't exist there, by design. we might not be able to handle performance/functionality issues.

You recently added paging support by using our search API, which isn't fully supported and won't be supported in the user-level API.
You are also supposed to change the VM MAC address search to be based on our search, rather than searching locally.

You can use that without paging and search and that will probably work (wasn't tested, though) , but as you rely more on the API in other use-cases in Foreman, you might run into using something that doesn't exist in the user level api, and you'll need to block stuff on your side, or suggest alternatives, if the user chose the filtered view.

In my opinion the requirement to be admin is reasonable. We require the same on the other side of the integration. Some API calls require you to be an admin.
Admin doesn't mean super user, btw, you just need to have one admin role assigned to one object, so in case you're managing things through Foreman - the ovirt admin should give you this minimum permission so you can work with it. Thus, I don't see this as a serious limitation.
How many customers asked for that?

Bottom line, as one of the owners of the oVirt API, my recommendation is not to support that, or support it but block functionality, and in that case we can't promise a blocked functionality will be added, and a performance issue at scale will be resolved.

ares commented:
Closing, we'll document what admin roles/permissions are required instead.
